source/common/access_log/access_log_formatter.cc

```diff
@@ -64,14 +64,55 @@ std::string FormatterImpl::format(const Http::HeaderMap& request_headers,
   std::string log_line;
   log_line.reserve(256);
 
-  for (const FormatterPtr& formatter : formatters_) {
-    log_line +=
-        formatter->format(request_headers, response_headers, response_trailers, stream_info);
+  for (const FormatterProviderPtr& provider : providers_) {
+    log_line += provider->format(request_headers, response_headers, response_trailers, stream_info);
   }
 
   return log_line;
 }
 
+JsonFormatterImpl::JsonFormatterImpl(std::unordered_map<std::string, std::string>& format_mapping) {
+  for (const auto& pair : format_mapping) {
+    auto providers = AccessLogFormatParser::parse(pair.second);
+    json_output_format_.emplace(pair.first, FormatterPtr{new FormatterImpl(pair.second)});
+  }
+}
+
+std::string JsonFormatterImpl::format(const Http::HeaderMap& request_headers,
+                                      const Http::HeaderMap& response_headers,
+                                      const Http::HeaderMap& response_trailers,
+                                      const StreamInfo::StreamInfo& stream_info) const {
+  const auto output_map =
+      this->toMap(request_headers, response_headers, response_trailers, stream_info);
```

Comment:
Nit: you should be able to just write `toMap` rather than `this->toMap`. Personally I would inline `toMap` since it is now private, but up to you.